File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public bool IsValidBST(TreeNode root) {
        return dfs(root, Int32.MinValue, Int32.MaxValue);
    }

    //helper function

    public bool dfs(TreeNode node,int minNodeVal,int  maxNodeVal){
        //base case
        if (node is null) return true;
        //check to see if it fits in the ideal range to be valid 
        if(!(node.val < maxNodeVal && node.val > minNodeVal)){
            return false;
        }

        return dfs(node.left,minNodeVal, node.val  ) && dfs(node.right, node.val, maxNodeVal);
    }
}

File: Data Structures & Algorithms/invert-a-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        #treat each node as its own tree beginning 
        if root is None:
            return
            
        #swap roots
        root.left,root.right = root.right, root.left

        self.invertTree(root.left)
        self.invertTree(root.right)

        return root
